source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,667 @@
+    if (vt->element_type()->AsBool())
+      return true;
+    else if (auto* vit = vt->element_type()->AsInteger()) {
+      if (vit->width() == 32) return true;
+    }
+  }
+  return false;
+}
+}
+
+ir::Instruction* FoldSpecConstantOpAndCompositePass::DoComponentWiseOperation(
+    ir::Module::inst_iterator* pos) {
+  const ir::Instruction* inst = &**pos;
+  const analysis::Type* result_type = GetType(inst);
+  SpvOp spec_opcode = static_cast<SpvOp>(inst->GetSingleWordInOperand(0));
+  // Check and collect operands.
+  std::vector<Const*> operands;
+  if (!TraverseConstOperands(*inst, [&operands, this](uint32_t id) {
+        if (Const* c = FindRecordedConst(id)) {
+          if (IsValidTypeForComponentWiseOperation(c->type())) {
+            operands.push_back(c);
+            return true;
+          }
+        }
+        return false;
+      })) {
+    return nullptr;
+  }
+
+  if (result_type->AsInteger() || result_type->AsBool()) {
+    // Scalar operation
+    uint32_t result_val = OperateScalars(spec_opcode, operands);
+    auto result_const = CreateConst(result_type, {result_val});
+    return BuildInstructionAndAddToModule(std::move(result_const), pos);
+  } else {
+    // Vector operation
+    const analysis::Type* element_type =
+        result_type->AsVector()->element_type();
+    auto new_vec_const =
+        std::unique_ptr<Const>(new VectorConst(result_type, element_type));
+    uint32_t num_dims = result_type->AsVector()->element_count();
+    std::vector<uint32_t> result_vec =
+        OperateVectors(spec_opcode, num_dims, operands);
+    for (uint32_t r : result_vec) {
+      auto rc = CreateConst(element_type, {r});
+      new_vec_const->AsVectorConst()->AppendComponent(rc.get());
+      BuildInstructionAndAddToModule(std::move(rc), pos);
+    }
+    return BuildInstructionAndAddToModule(std::move(new_vec_const), pos);
+  }
+}
+
+ir::Instruction*
+FoldSpecConstantOpAndCompositePass::BuildInstructionAndAddToModule(
+    std::unique_ptr<Const> c, ir::Module::inst_iterator* pos) {
+  Const* new_const = c.get();
+  uint32_t new_id = ++max_id_;
+  const_val_to_id_[c.get()] = new_id;
+  id_to_const_val_[new_id] = std::move(c);
+  auto new_inst = CreateInstruction(new_id, new_const);
+  if (!new_inst) return nullptr;
+  auto* new_inst_ptr = new_inst.get();
+  *pos = pos->InsertBefore(std::move(new_inst));
+  (*pos)++;
+  def_use_mgr_->AnalyzeInstDefUse(new_inst_ptr);
+  return new_inst_ptr;
+}
+
+std::unique_ptr<Const> FoldSpecConstantOpAndCompositePass::CreateConstFromInst(
+    ir::Instruction* inst) {
+  std::vector<uint32_t> literal_words_or_ids;
+  std::unique_ptr<Const> new_const;
+  for (uint32_t i = 0; i < inst->NumInOperands(); i++) {
+    literal_words_or_ids.insert(literal_words_or_ids.end(),
+                                inst->GetInOperand(i).words.begin(),
+                                inst->GetInOperand(i).words.end());
+  }
+  switch (inst->opcode()) {
+    case SpvOp::SpvOpConstantTrue:
```

Comment:
I'd suggest to put some comments here to say that `OpConstant{True|False}` embedded the value in the opcode. So they are not handled in the above loop for in operands, while the following opcodes are handled above. Otherwise, it reads a little bit strange here.